embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/resurrection/PayloadResurrectionServiceImpl.kt

```diff
@@ -36,68 +38,88 @@ internal class PayloadResurrectionServiceImpl(
 
     override fun resurrectOldPayloads(nativeCrashServiceProvider: Provider<NativeCrashService?>) {
         val nativeCrashService = nativeCrashServiceProvider()
-        val nativeCrashes = nativeCrashService
-            ?.getNativeCrashes()
-            ?.associateBy { it.sessionId }
-            ?.apply {
-                values.forEach { nativeCrash ->
-                    nativeCrashService.sendNativeCrash(nativeCrash)
-                }
-            } ?: emptyMap()
+        val undeliveredPayloads = cacheStorageService.getUndeliveredPayloads()
+        val nativeCrashes = nativeCrashService?.getNativeCrashes()?.associateBy { it.sessionId } ?: emptyMap()
 
-        cacheStorageService
-            .getUndeliveredPayloads()
-            .forEach { deadProcessPayloadMetadata ->
-                deadProcessPayloadMetadata.processUndeliveredPayload(nativeCrashes::get)
-            }
+        processUndeliveredPayload(undeliveredPayloads, nativeCrashes, nativeCrashService)
         nativeCrashService?.deleteAllNativeCrashes()
     }
 
     /**
      * Load and modify the given incomplete payload envelope and send the result to the [IntakeService] for delivery.
      * Resurrected payloads sent to the [IntakeService] will be deleted.
      */
-    private fun StoredTelemetryMetadata.processUndeliveredPayload(nativeCrashProvider: (String) -> NativeCrashData?) {
-        val result = runCatching {
-            val resurrectedPayload = when (envelopeType) {
-                SupportedEnvelopeType.SESSION -> {
-                    val deadSession = serializer.fromJson<Envelope<SessionPayload>>(
-                        inputStream = GZIPInputStream(cacheStorageService.loadPayloadAsStream(this)),
-                        type = envelopeType.serializedType
-                    )
+    private fun processUndeliveredPayload(
```

Comment:
This function is fairly long with lots of indentation & that makes it hard to read IMO. I'd suggest breaking it into 2-3 smaller functions